File: ts/project/.sdk/tm/elixir/lib/projectname/feature/rbac.ex
# ProjectName SDK rbac feature
#
# Client-side role/permission enforcement. Before an operation resolves its
# endpoint, the required permission for that entity+operation is checked
# against the permissions the client holds; a disallowed call is
# short-circuited with an `rbac_denied` error placed in ctx.out["point"]
# (the pipeline's make_point surfaces it) and never touches the network.
# Required permissions come from `rules` (keyed by `<entity>.<op>`, `<op>`,
# or `*`); the default when no rule matches is controlled by `deny`
# (default: allow when unspecified). Held permissions are the `permissions`
# list (a `*` grants everything).

defmodule ProjectName.Feature.Rbac do
  alias Voxgig.Struct, as: S
  alias ProjectName.{Feature, Context}

  def new do
    f = Feature.base("rbac")
    Feature.install(f, "init", fn ctx, opts -> init(f, ctx, opts) end)
    Feature.install(f, "PrePoint", fn ctx -> pre_point(f, ctx) end)
    f
  end

  def init(f, ctx, options) do
    Feature.init_common(f, ctx, options)

    opts = Feature.opts(f)
    granted = S.jm([])
    perms = S.getprop(opts, "permissions")

    if S.islist(perms) do
      n = S.size(perms)

      if n > 0 do
        Enum.each(0..(n - 1), fn i ->
          perm = S.getelem(perms, i)
          S.setprop(granted, perm, true)
        end)
      end
    end

    S.setprop(f, "granted", granted)
    nil
  end

  defp pre_point(f, ctx) do
    if Feature.active?(f) do
      required = required(f, ctx)

      cond do
        required == nil ->
          # No rule: honour the default policy.
          if S.getprop(Feature.opts(f), "deny") == true do
            reject(f, ctx, "<default-deny>")
          end

          nil

        true ->
          granted = S.getprop(f, "granted")

          if S.getprop(granted, "*") == true or S.getprop(granted, required) == true do
            track(f, ctx, required, true)
          else
            reject(f, ctx, required)
          end
      end
    end

    nil
  end

  defp required(f, ctx) do
    rules = S.getprop(Feature.opts(f), "rules")
    rules = if S.ismap(rules), do: rules, else: S.jm([])

    entity = entity_of(ctx)
    opname = opname_of(ctx)
    key = entity <> "." <> opname

    cond do
      S.getprop(rules, key) != nil -> S.getprop(rules, key)
      S.getprop(rules, opname) != nil -> S.getprop(rules, opname)
      S.getprop(rules, "*") != nil -> S.getprop(rules, "*")
      true -> nil
    end
  end

  defp entity_of(ctx) do
    name = Context.entity_name(S.getprop(ctx, "entity"))
    name = if name == "_", do: "", else: name

    if name == "" do
      op = S.getprop(ctx, "op")
      oe = if op != nil, do: S.getprop(op, "entity"), else: nil
      if is_binary(oe), do: oe, else: ""
    else
      name
    end
  end

  defp opname_of(ctx) do
    op = S.getprop(ctx, "op")
    nm = if op != nil, do: S.getprop(op, "name"), else: nil
    if is_binary(nm), do: nm, else: ""
  end

  defp reject(f, ctx, required) do
    track(f, ctx, required, false)

    op = S.getprop(ctx, "op")
    nm = if op != nil, do: S.getprop(op, "name"), else: nil
    opname = if is_binary(nm) and nm != "", do: nm, else: "?"

    err =
      Context.make_error(
        ctx,
        "rbac_denied",
        "Permission \"" <> to_string(required) <> "\" required for operation \"" <> opname <> "\""
      )

    # Short-circuit endpoint resolution; the pipeline surfaces this error.
    S.setprop(S.getprop(ctx, "out"), "point", err)
    err
  end

  defp track(f, ctx, required, allowed) do
    client = S.getprop(f, "client")
    t = Feature.track_node(client, "_rbac", S.jm(["allowed", 0, "denied", 0, "last", nil]))

    key = if allowed, do: "allowed", else: "denied"
    S.setprop(t, key, S.getprop(t, key) + 1)

    op = S.getprop(ctx, "op")
    opname = if op != nil, do: S.getprop(op, "name"), else: nil
    S.setprop(t, "last", S.jm(["required", required, "allowed", allowed, "op", opname]))
    nil
  end
end

File: ts/project/.sdk/tm/zig/feature/rbac.zig
// Client-side role/permission enforcement (mirrors go feature/rbac_feature.go
// / rust feature/rbac.rs). Before an operation resolves its endpoint, the
// required permission for that entity+operation is checked against the held
// permissions; a disallowed call is short-circuited with an `rbac_denied`
// error stored on ctx.out["point"] (which make_point surfaces — gotcha #2) and
// never touches the network.

const std = @import("std");
const h = @import("../core/helpers.zig");
const types = @import("../core/types.zig");
const sup = @import("support.zig");

const Value = h.Value;
const Context = types.Context;
const Feature = types.Feature;
const OutVal = types.OutVal;

pub const RbacFeature = struct {
    name: []const u8 = "rbac",
    active: bool = true,
    add_opts: Value = .{ .null = {} },
    options: Value = .{ .null = {} },
    granted: std.StringHashMap(bool),

    allowed: i64 = 0,
    denied: i64 = 0,
    last: Value = .{ .null = {} },

    pub fn make() Feature {
        const self = h.A().create(RbacFeature) catch unreachable;
        self.* = .{ .granted = std.StringHashMap(bool).init(h.A()) };
        return .{ .ptr = @ptrCast(self), .vtable = &vtable };
    }

    fn self_of(p: *anyopaque) *RbacFeature {
        return @ptrCast(@alignCast(p));
    }

    fn required(self: *RbacFeature, ctx: *Context) ?[]const u8 {
        const rules = sup.fopt_map(self.options, "rules");
        if (rules != .object) return null;

        const entity: []const u8 = if (ctx.entity) |e| e.get_name() else ctx.op.entity;
        const opname = ctx.op.name;

        const k0 = std.fmt.allocPrint(h.A(), "{s}.{s}", .{ entity, opname }) catch opname;
        const keys = [_][]const u8{ k0, opname, "*" };
        for (keys) |key| {
            if (h.get_str(rules, key)) |r| return r;
        }
        return null;
    }

    fn track(self: *RbacFeature, ctx: *Context, req: []const u8, allowed: bool) void {
        if (allowed) {
            self.allowed += 1;
        } else {
            self.denied += 1;
        }
        const opname = ctx.op.name;
        const last = h.omap();
        h.setp(last, "required", h.vstr(req));
        h.setp(last, "allowed", h.vbool(allowed));
        h.setp(last, "op", h.vstr(opname));
        self.last = last;
    }

    fn reject(self: *RbacFeature, ctx: *Context, req: []const u8) void {
        self.track(ctx, req, false);
        const opname0 = ctx.op.name;
        const opname = if (opname0.len == 0) "?" else opname0;
        const msg = std.fmt.allocPrint(h.A(), "Permission \"{s}\" required for operation \"{s}\"", .{ req, opname }) catch "rbac denied";
        const e = ctx.make_error("rbac_denied", msg);
        // Short-circuit endpoint resolution; make_point surfaces this error.
        ctx.out_set("point", OutVal{ .err = e });
    }

    fn pre_point(self: *RbacFeature, ctx: *Context) void {
        if (!self.active) return;

        const req = self.required(ctx) orelse {
            // No rule: honour the default policy.
            if (sup.fopt_bool(self.options, "deny", false)) {
                self.reject(ctx, "<default-deny>");
            }
            return;
        };

        if ((self.granted.get("*") orelse false) or (self.granted.get(req) orelse false)) {
            self.track(ctx, req, true);
            return;
        }
        self.reject(ctx, req);
    }

    fn vname(p: *anyopaque) []const u8 {
        return self_of(p).name;
    }
    fn vactive(p: *anyopaque) bool {
        return self_of(p).active;
    }
    fn vaddopts(p: *anyopaque) Value {
        return self_of(p).add_opts;
    }
    fn vinit(p: *anyopaque, ctx: *Context, options: Value) void {
        _ = ctx;
        const self = self_of(p);
        self.options = options;
        self.active = sup.fopt_bool(options, "active", false);
        self.granted = std.StringHashMap(bool).init(h.A());
        if (sup.fopt_str_list(options, "permissions")) |perms| {
            for (perms) |perm| self.granted.put(perm, true) catch {};
        }
    }
    fn vdispatch(p: *anyopaque, name: []const u8, ctx: *Context) void {
        if (std.mem.eql(u8, name, "PrePoint")) self_of(p).pre_point(ctx);
    }

    const vtable = Feature.VTable{
        .name = vname,
        .active = vactive,
        .add_options = vaddopts,
        .init = vinit,
        .dispatch = vdispatch,
    };
};
